File: packages/primitives/src/Signature.ts
import { Schema, Effect } from "effect";
import { Bytes } from "ox";
import { type B256, B256 as B256Schema } from "./B256.js";
import { type Address } from "./Address.js";
import { fromHex } from "viem/utils";
import { toBytes } from "viem/utils";

/**
 * An Ethereum ECDSA signature
 */
export interface Signature {
  r: B256;
  s: B256;
  v: number; // recid (recovery identifier)
  yParity: 0 | 1; // Recovery identifier in compact form
}

/**
 * Schema for validating Signature objects
 */
export const Signature = Schema.Struct({
  r: B256Schema,
  s: B256Schema,
  v: Schema.Number,
  yParity: Schema.Union(Schema.Literal(0), Schema.Literal(1))
});

/**
 * Creates a Signature from raw components
 * @param r - The r component of the signature.
 * @param s - The s component of the signature.
 * @param v - The v component of the signature (recid).
 */
export const fromComponents = (
  r: B256,
  s: B256,
  v: number
): Effect.Effect<Signature, Error> =>
  Effect.gen(function*(_) {
    const yParity = v % 2 === 0 ? 0 : 1;
    
    return {
      r,
      s,
      v,
      yParity: yParity as (0 | 1)
    };
  });

/**
 * Creates a Signature from a raw byte array
 * @param bytes - The raw signature bytes.
 */
export const fromRaw = (bytes: Uint8Array): Effect.Effect<Signature, Error> =>
  Effect.gen(function*(_) {
    if (bytes.length !== 65) {
      return yield* _(Effect.fail(new Error("Signature must be 65 bytes")));
    }
    
    const r = bytes.slice(0, 32);
    const s = bytes.slice(32, 64);
    const v = bytes[64];
    
    return yield* _(fromComponents(r, s, v));
  });

/**
 * Creates a Signature from a hex string
 * @param hex - The hex string.
 */
export const fromHexString = (hex: string): Effect.Effect<Signature, Error> =>
  Effect.gen(function*(_) {
    const hexWithPrefix = hex.startsWith("0x") ? hex : `0x${hex}`;
    
    if (hexWithPrefix.length !== 132) { // 0x + 65 bytes * 2 hex chars
      return yield* _(Effect.fail(new Error("Signature hex must be 130 characters (65 bytes)")));
    }
    
    const bytes = fromHex(hexWithPrefix);
    return yield* _(fromRaw(bytes));
  });

/**
 * Creates a Signature from ERC-2098 compact format (64 bytes)
 * @param bytes - The compact signature bytes.
 */
export const fromErc2098 = (bytes: Uint8Array): Effect.Effect<Signature, Error> =>
  Effect.gen(function*(_) {
    if (bytes.length !== 64) {
      return yield* _(Effect.fail(new Error("ERC-2098 signature must be 64 bytes")));
    }
    
    const r = bytes.slice(0, 32);
    const s = bytes.slice(32, 64);
    
    // Extract yParity from the highest bit of s
    const yParity = (s[0] & 0x80) ? 1 : 0;
    
    // Clear the highest bit of s
    const sCopy = new Uint8Array(s);
    sCopy[0] = sCopy[0] & 0x7f;
    
    // Calculate v
    const v = yParity + 27;
    
    return {
      r,
      s: sCopy,
      v,
      yParity: yParity as (0 | 1)
    };
  });

/**
 * Converts a Signature to raw bytes
 * @param signature - The Signature.
 */
export const toRaw = (signature: Signature): Effect.Effect<Uint8Array> =>
  Effect.sync(() => {
    const result = new Uint8Array(65);
    result.set(signature.r, 0);
    result.set(signature.s, 32);
    result[64] = signature.v;
    return result;
  });

/**
 * Converts a Signature to a hex string
 * @param signature - The Signature.
 */
export const toHex = (signature: Signature): Effect.Effect<string> =>
  Effect.gen(function*(_) {
    const bytes = yield* _(toRaw(signature));
    return Bytes.toHex(bytes);
  });

/**
 * Converts a Signature to ERC-2098 compact format
 * @param signature - The Signature.
 */
export const toErc2098 = (signature: Signature): Effect.Effect<Uint8Array> =>
  Effect.sync(() => {
    const result = new Uint8Array(64);
    result.set(signature.r, 0);
    
    // Create a copy of s
    const sCopy = new Uint8Array(signature.s);
    
    // Set the highest bit of s according to yParity
    if (signature.yParity === 1) {
      sCopy[0] = sCopy[0] | 0x80;
    } else {
      sCopy[0] = sCopy[0] & 0x7f;
    }
    
    result.set(sCopy, 32);
    return result;
  });

/**
 * Recovers the signer's address from a signature and message 
 * (Note: Currently a placeholder as we need viem account interface)
 * @param signature - The Signature.
 * @param message - The original message that was signed.
 */
export const recoverAddressFromMsg = (
  signature: Signature,
  message: Uint8Array
): Effect.Effect<Address, Error> =>
  Effect.fail(new Error("Not implemented yet - needs viem account interface"))

/**
 * Recovers the signer's address from a signature and message hash
 * (Note: Currently a placeholder as we need viem account interface)
 * @param signature - The Signature.
 * @param messageHash - The hash of the message that was signed.
 */
export const recoverAddressFromHash = (
  signature: Signature,
  messageHash: B256
): Effect.Effect<Address, Error> =>
  Effect.fail(new Error("Not implemented yet - needs viem account interface"))

/**
 * Normalizes the 's' value of a signature according to EIP-2
 * @param signature - The Signature.
 */
export const normalizeS = (signature: Signature): Effect.Effect<Signature, Error> =>
  Effect.gen(function*(_) {
    // Define the secp256k1 curve order / 2
    const SECP256K1_N_DIV_2 = 0x7fffffffffffffffffffffffffffffff5d576e7357a4501ddfe92f46681b20a0n;
    
    // Convert s to bigint
    const s = Bytes.toBigInt(signature.s);
    
    // If s is already normalized, return the signature as is
    if (s <= SECP256K1_N_DIV_2) {
      return signature;
    }
    
    // Otherwise, calculate N - s
    const SECP256K1_N = 0xfffffffffffffffffffffffffffffffebaaedce6af48a03bbfd25e8cd0364141n;
    const normalizedS = SECP256K1_N - s;
    
    // Convert back to bytes
    const normalizedSBytes = fromHex(`0x${normalizedS.toString(16).padStart(64, '0')}`);
    
    // Decode into a B256
    const normalizedSB256 = yield* _(Schema.decode(B256Schema)(normalizedSBytes));
    
    // Flip v/yParity
    const newYParity = signature.yParity === 0 ? 1 : 0;
    const newV = signature.v % 2 === 0 ? signature.v + 1 : signature.v - 1;
    
    // Return the normalized signature
    return {
      r: signature.r,
      s: normalizedSB256,
      v: newV,
      yParity: newYParity
    };
  });

File: src/signature/signature.zig
const std = @import("std");
const error = @import("error.zig");
const utils = @import("utils.zig");

/// An Ethereum ECDSA signature.
/// Represents a signature with r, s components and y_parity boolean.
pub const Signature = struct {
    /// Whether the Y coordinate of the curve point is odd
    y_parity: bool,
    /// The r component of the signature (32 bytes)
    r: [32]u8,
    /// The s component of the signature (32 bytes)
    s: [32]u8,

    /// Creates a signature from r, s components and y_parity flag
    ///
    /// r: The r component of the signature (32 bytes)
    /// s: The s component of the signature (32 bytes)
    /// y_parity: Whether the Y coordinate is odd
    ///
    /// Returns an error if s is not in the lower half of the curve order
    pub fn from_rs_and_y_parity(r: [32]u8, s: [32]u8, y_parity: bool) !Signature {
        // Pseudocode:
        // 1. Check that s is in the lower half of the curve order
        // 2. If s is not in the lower half, return error.InvalidS
        // 3. Return a new Signature with the provided components
        @compileError("Not implemented");
    }
    
    /// Parse a signature from a 65-byte raw format (r[32] + s[32] + v[1])
    pub fn from_raw(bytes: []const u8) !Signature {
        // Pseudocode:
        // 1. Verify that bytes is exactly 65 bytes long
        // 2. Extract r (first 32 bytes)
        // 3. Extract s (next 32 bytes)
        // 4. Extract v (last byte) and convert to y_parity boolean
        // 5. Call fromRsAndYParity with extracted values
        @compileError("Not implemented");
    }
    
    /// Parse a signature from a hex string
    pub fn from_hex(hex: []const u8) !Signature {
        // Pseudocode:
        // 1. Verify hex is the correct length (with or without 0x prefix)
        // 2. Convert hex to bytes
        // 3. Call from_raw with converted bytes
        @compileError("Not implemented");
    }
    
    /// Converts the signature to a 65-byte array in the format r[32] + s[32] + v[1]
    pub fn as_bytes(self: Signature) [65]u8 {
        // Pseudocode:
        // 1. Create a 65-byte array
        // 2. Copy r into first 32 bytes
        // 3. Copy s into next 32 bytes
        // 4. Set last byte to 0 or 1 based on y_parity
        // 5. Return the array
        @compileError("Not implemented");
    }
    
    /// Converts the signature to a hexadecimal string with 0x prefix
    pub fn to_hex(self: Signature, allocator: std.mem.Allocator) ![]u8 {
        // Pseudocode:
        // 1. Get raw bytes using as_bytes()
        // 2. Allocate buffer for hex string (2 chars per byte + 2 for 0x)
        // 3. Write 0x prefix
        // 4. Convert bytes to hex characters
        // 5. Return the resulting string
        @compileError("Not implemented");
    }
    
    /// Verifies that the signature is in canonical form
    /// (s is in the lower half of the curve order)
    pub fn is_valid(self: Signature) bool {
        // Pseudocode:
        // 1. Check that s is in the lower half of the curve order
        // 2. Return true if it is, false otherwise
        @compileError("Not implemented");
    }
    
    /// Recovers the public key that was used to create this signature
    /// for the given message hash
    pub fn recover_public_key(self: Signature, message_hash: [32]u8) ![65]u8 {
        // Pseudocode:
        // 1. Use secp256k1 recovery to get public key from signature and message hash
        // 2. Format public key as uncompressed (65 bytes)
        // 3. Return the public key bytes
        @compileError("Not implemented");
    }
    
    /// Recovers the Ethereum address that was used to create this signature
    /// for the given message hash
    pub fn recover_address(self: Signature, message_hash: [32]u8) ![20]u8 {
        // Pseudocode:
        // 1. Recover public key using recoverPublicKey
        // 2. Take keccak256 hash of the public key (excluding the first byte)
        // 3. Extract the last 20 bytes of the hash as the address
        // 4. Return the address
        @compileError("Not implemented");
    }
    
    /// Create an Ethereum-prefixed hash from a message
    /// Uses the "\x19Ethereum Signed Message:\n" prefix
    pub fn hash_message(message: []const u8) [32]u8 {
        // Pseudocode:
        // 1. Create the prefix string "\x19Ethereum Signed Message:\n" + message.length
        // 2. Hash prefix + message using keccak256
        // 3. Return the resulting hash
        @compileError("Not implemented");
    }
};